scripts/histmakers/mw_with_mu_eta_pt.py

```diff
@@ -290,10 +290,11 @@ def build_graph(df, dataset):
     df = muon_selections.veto_electrons(df)
     df = muon_selections.apply_met_filters(df)
     if args.makeMCefficiency:
-        df = df.Define("GoodTrigObjs", "wrem::goodMuonTriggerCandidate(TrigObj_id,TrigObj_filterBits)")
-        df = df.Define("passTrigger","(HLT_IsoTkMu24 || HLT_IsoMu24) && wrem::hasTriggerMatch(goodMuons_eta0,goodMuons_phi0,TrigObj_eta[GoodTrigObjs],TrigObj_phi[GoodTrigObjs])")
+        df = df.Define("GoodTrigObjs", f"wrem::goodMuonTriggerCandidate(TrigObj_id,TrigObj_filterBits, \"{era}\")")
```

Comment:
The trigger that is used is known at runtime, I suggest to solve this a bit different. For example, defining two separate functions for 2016 and 2017/2018.

Reply:
Hi David, thanks for the suggestion. Originally we were actually doing as you say, but some functions are used internally by others, for example here (where by the way you may have spotted a bug) [0]. So we might still have to allow for an "era" flag in the function definition sometimes. At that point we thought it was simpler to have it wherever it could be used, keeping 2016 as default to minimally affect other pieces of code. 

In case you were also concerned about the usage of strings for the era, we might define an enum class paired with a dictionary to map integers into string for printing purpose, similarly to what is currently done for the analysis type, so to pass an integer type to all functions.

[0] https://github.com/WMass/WRemnants/blob/a9276858cf93cbd97188ac11cabd8173dc28e7ea/wremnants/muon_selections.py#L100C5-L109

Reply:
I think it's fine to have the era flag in python functions, I would try to avoid it in the c++ function because those are then used in the computational graph. Another possibility is to write c++ template functions, but two different c++ functions would be simpler in my opinion.

Reply:
I would not use distinct functions for each era because we would end up with duplicated code.

It _does_ make sense to have a single function of `era` (and other parameters if needed) that assembles and outputs the trigger bit string. 

I would keep things as is right now because it's only a single if statement. If the trigger definition becomes more complex then we can replace it with a function.

Reply:
To avoid duplication of code you can make it a template function, something like
```
enum class Era {
    Era_2016PostVFP,
    // Add other eras as needed
};

template <Era era>
Vec_i goodMuonTriggerCandidate(const Vec_i& TrigObj_id, const Vec_f& TrigObj_pt, const Vec_f& TrigObj_l1pt, const Vec_f& TrigObj_l2pt, const Vec_i& TrigObj_filterBits) {
    Vec_i res(TrigObj_id.size(), 0);
    if (TrigObj_id[i]  != 13 ) continue;
    if (TrigObj_pt[i]   < 24.) continue;
    if (TrigObj_l1pt[i] < 22.) continue;
    if constexpr (era == Era::Era_2016PostVFP){
      if (! ((TrigObj_filterBits[i] & 8) || (TrigObj_l2pt[i] > 10. && (TrigObj_filterBits[i] & 2))))) continue;
    }
    if constexpr (era != Era::Era_2016PostVFP){
      if (! (TrigObj_l2pt[i] > 10. && (TrigObj_filterBits[i] & 2)))) continue;
    }
    res[i] = 1;
   }
   return res;
}
```
and call the function as 
```
goodMuonTriggerCandidate<Era::Era_2016PostVFP>(TrigObj_id, TrigObj_pt, TrigObj_l1pt, TrigObj_l2pt, TrigObj_filterBits);
```
This make the graph indepent of era avoids passing the era string into the function for every single event

Reply:
I would be in favour of this (either the template or the duplication), mainly because I suspect using string comparison per-event can be quite slow (and unnecessary).
Since currently we foresee only two cases, instead of two independent "if"s we can use "if .. else".

Reply:
I also agree. I'll work on this.